server/controllers/article.js

```diff
@@ -5,6 +5,155 @@ import paginateArticle from '../utils/articlesPaginate';
+    return tags;
+  }
+
+  /**
+   * @param {list} tag
+   * @param {object} res
+   * @param {object} next
+   * @return {object} jsonResponse
+   */
+  static byTags(tag, res, next) {
+    // trats each word as separate tag
+    const tags = Search.tokenize(tag);
+
+    Tag.findAll({
+      where: {
+        $or: [
+          {
+            tag: tags
+          }
+        ]
+      },
+      include:
+        [{
+          model: Article,
+          as: 'articles',
+          through: {
+            attributes: []
+          },
+          include: [{
+            model: User,
+            as: 'author',
+            attributes: ['username', 'email', 'bio', 'image']
+          },
+          {
+            model: Tag,
+            as: 'tags',
+            attributes: ['tag'],
+            through: {
+              attributes: [],
+            }
+          }
+          ]
+        }
+        ],
+      attributes: ['id', 'tag']
+    }).then(articles => res.status(200).json({
+      status: 'success',
+      articles
+    })).catch(next);
+  }
+
+  /**
+   * @param {list} keywords
+   * @param {object} res
+   * @param {object} next
+   * @return {object} jsonResponse
+   */
+  static byKeywords(keywords, res, next) {
+    Article.findAll({
+      where: {
+        $or: [
+          {
+            title: { like: `%${keywords}%` }
+          },
+          {
+            description: { like: `%${keywords}%` }
+          }
+        ]
+      },
+      include: [{
+        model: User,
+        as: 'author',
+        attributes: ['username', 'email', 'bio', 'image']
+      },
+      {
+        model: Tag,
+        as: 'tags',
+        attributes: ['tag'],
+        through: {
+          attributes: [],
+        }
+      }
+      ]
+    }).then(articles => res.status(200).json({
+      status: 'success',
+      articles
+    })).catch(next);
+  }
+
+  /**
+   * @param {String} author
+   * @param {object} res
+   * @param {object} next
+   * @return {object} jsonResponse
+   */
+  static byAuthor(author, res, next) {
+    User.findAll({
+      where: {
+        $or: [
+          {
+            username: { like: `%${author}%` }
+          },
+          {
+            firstName: { like: `%${author}%` }
+          },
+          {
+            lastName: { like: `%${author}%` }
+          }
+        ]
+      },
+      include:
+          [{
+            model: Article,
+            as: 'authored',
+
+            include: [{
+              model: Tag,
+              as: 'tags',
+              attributes: ['tag'],
+              through: {
+                attributes: [],
+              }
+            },
+            {
+              model: User,
+              as: 'author',
+              attributes: ['username', 'email', 'bio', 'image']
+            }
+            ]
+          }],
+      attributes: ['id', 'username']
+    }).then(articles => res.status(200).json({
+      status: 'success',
+      articles
+    })).catch(next);
+  }
+}
+
```

Comment:
Nice work on this implementation of search functionality. However, in lines `11-157`, I think you can make the controllers leaner if you extract the Search class into its' own javascript file and then import its' static methods for use in the `articleController`.

I think, your line of thought was geared toward something like private methods, but I think implementing those methods from the search class as helpers/middleware would provide better separation of concerns.